_dev/sources/addin.ts

```diff
@@ -364,27 +381,57 @@ class Addin {
         }).finally(() => this.toggleWaiting());
     }
 
+    saveChanges = () => {
+        this.render();
+        (<HTMLInputElement>this.exportBtn).disabled = false;
+    }
+
+    checkBoxValueChanged = () => {
+        (<HTMLInputElement>this.exportBtn).disabled = true;
+    }
+
     render () {
+        this.data.users = [];
+        this.data.zones = [];
+        this.setAddinsToNull();
+        //wire up the dom
         let mapMessageTemplate: string = document.getElementById("mapMessageTemplate").innerHTML,
             groupsBlock: HTMLElement = document.getElementById("exportedGroups"),
             securityClearancesBlock: HTMLElement = document.getElementById("exportedSecurityClearances"),
             rulesBlock: HTMLElement = document.getElementById("exportedRules"),
             reportsBlock: HTMLElement = document.getElementById("exportedReports"),
             dashboardsBlock: HTMLElement = document.getElementById("exportedDashboards"),
             addinsBlock: HTMLElement = document.getElementById("exportedAddins"),
-            thisAddinBlock: HTMLElement = document.getElementById("includeThisAddin"),
-            thisAddinIncludedCheckbox: HTMLElement = document.querySelector("#includeThisAddin > input"),
-            mapBlockDescription: HTMLElement = document.querySelector("#exportedMap > .description");
+            // exportAllAddinsCheckbox: HTMLInputElement = document.getElementById("export_all_addins_checkbox") as HTMLInputElement,
+            // thisAddinBlock: HTMLElement = document.getElementById("includeThisAddin"),
+            // thisAddinIncludedCheckbox: HTMLElement = document.querySelector("#includeThisAddin > input"),
+            mapBlockDescription: HTMLElement = document.querySelector("#exportedMap > .description"),
+            usersBlock: HTMLElement = document.getElementById("exportedUsers"),
+            // exportAllUsersCheckbox: HTMLInputElement = document.getElementById("export_all_users_checkbox") as HTMLInputElement,
+            zonesBlock: HTMLElement = document.getElementById("exportedZones");
+            // exportAllZonesCheckbox: HTMLInputElement = document.getElementById("export_all_zones_checkbox") as HTMLInputElement;
+        //wire up the export button event
         this.exportBtn.addEventListener("click", this.exportData, false);
-        thisAddinIncludedCheckbox.addEventListener("change", this.toggleThisAddinIncluded, false);
+        this.saveBtn.addEventListener("click", this.saveChanges, false);
```

Comment:
You are requesting zones and users in any case. It would be great to remove this button. It can be confusing for users.

Reply:
Nick it is linked to the checkboxes in the page. if they change the script needs to be re-executed. have to tried running the application?

Reply:
Yes. I ran the addin and it seemed confising to me. You recall all entities on pressing "Save". It doesn't depend on the state of checkboxes:
```
this.userBuilder.fetch(),
this.zoneBuilder.fetch()
```
You already know the quantity of entities for both checkbox states.
Another option is to leave "Save" button but don't fetch all zones and users when corresponding checkboxes are unchecked. The final choice depends on the expected total quantities of these entities. If you are going to export a lot of users or zones (hundreds or thousands) then it makes sense to leave "Save" button.

Reply:
we are talking about 2 states here:
1. only dependent users/zones
2. all users/zones
Default = 1
If you select the check box all users/zones are exported and the text in the page reflects the number.
I don't understand what you mean sorry. The quantity changes depending on what you would like...

Reply:
I meant that after loading the add-in you already fetched all Users and zones (lines 433, 434). So you know the total quantity of Users and Zones. After combineDependencies method (line 452) you know the total quantity of Users and Zones that are dependencies of other entities (like Rules and Reports). What you need to do is store them and update the message after changing the corresponding checkbox state. E.g. after unchecking "Export all users" checkbox you can show the quantity of users you received on line 452.

Reply:
ok so are you saying to leave the "Export all addins" & "Export all zones" check boxes and when they change just update the number without a Save/Refresh?

Reply:
Yes. That's what I wanted to say.

Reply:
I am not sure how to handle this. Example, If all users are selected and the user then de-selects this option I will need to render again is that correct? This to resolve the changed dependencies. Or are the dependencies resolved on the submit again? So on each check box change a render is executed. This is what I was trying to avoid with the Save button.

Reply:
Yes. You are right. Now you need to call "render" method again to get correct numbers. You should refactor it. It should be split to render() and e.g. getData() methods. Because now you add event listeners to buttons and checkboxes every time you call render(). But you don't remove those that were added previous time. That will lead to memory leaks.

Reply:
Please add removeEventListener for checkboxes in unload() method

Reply:
as agreed I have left the save button in place due to render method. To be reviewed in a later release.

Reply:
You forgot:
"You should refactor it. It should be split to render() and e.g. getData() methods. Because now you add event listeners to buttons and checkboxes every time you call render(). But you don't remove those that were added previous time. That will lead to memory leaks."